### PR TITLE
[FEATURE] Mesh 1D Renderer 

### PR DESCRIPTION
**Pull request by PeterPetrik**
Update to [MDAL 0.5.90](https://github.com/lutraconsulting/MDAL/releases/tag/0.5.90)

Initial implementation of [Support of 1D elements in Mesh Layers](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/164)

1D Meshes consist of vertices and edges. An edge connects two vertices and can have assigned data (scalars or vectors) on it. The 1D mesh network can be for example used for modelling of an urban drainage system.

There are new options to display 1D mesh (edges) in the mesh frame tab
Screenshot 2020-03-04 at 14 36 02

Also for data on 1D mesh (edges or vertices) there are new options in the contours tab

Screenshot 2020-03-04 at 14 35 34

Note that not all functionality available for 2D meshes are available, notably
- Support for Mesh Calculator for 1D meshes
- 3D rendering of 1D meshes 
- Identify and Plots in Crayfish (will be added later in the following PR)

Some functions are not necessary for 1D meshes, especially
- Export of mesh contours

### CODE REVIEW

File: src/core/mesh/qgsmeshlayerrenderer.cpp [hunk 1]
```diff
@@ -345,15 +413,223 @@ void QgsMeshLayerRenderer::renderScalarDataset()
...
+{
+  int r, g, b, a;
+  if ( shader->shade( val, &r, &g, &b, &a ) )
+  {
+    return QColor( r, g, b, a );
+  }
+  return QColor();
+}
+
+QgsPointXY QgsMeshLayerRenderer::fractionPoint( const QgsPointXY &p1, const QgsPointXY &p2, double fraction, double overlayMagniture ) const
```
Comment (wonder-sk) [inline src/core/mesh/qgsmeshlayerrenderer.cpp]: Can we use QgsGeometryUtils::interpolatePointOnLine() instead?

the `overlayMagniture` argument seems to be always called with 0, do we need it?

File: src/core/mesh/qgsmeshrenderersettings.cpp [hunk 2]
```diff
@@ -63,8 +74,8 @@ void QgsMeshRendererMeshSettings::readXml( const QDomElement &elem )
   mEnabled = elem.attribute( QStringLiteral( "enabled" ) ).toInt();
   mLineWidth = elem.attribute( QStringLiteral( "line-width" ) ).toDouble();
   mColor = QgsSymbolLayerUtils::decodeColor( elem.attribute( QStringLiteral( "color" ) ) );
+  mLineWidthUnit = QgsUnitTypes::decodeRenderUnit( QStringLiteral( "line-width-unit" ) );
```
Comment (wonder-sk) [inline src/core/mesh/qgsmeshrenderersettings.cpp]: missing elem.attribute( ... )

File: src/core/mesh/qgsmeshrenderersettings.cpp [hunk 3]
```diff
@@ -129,6 +142,9 @@ void QgsMeshRendererScalarSettings::readXml( const QDomElement &elem )
   mClassificationMinimum = elem.attribute( QStringLiteral( "min-val" ) ).toDouble();
   mClassificationMaximum = elem.attribute( QStringLiteral( "max-val" ) ).toDouble();
   mOpacity = elem.attribute( QStringLiteral( "opacity" ) ).toDouble();
+  mEdgeWidth = elem.attribute( QStringLiteral( "edge-width" ) ).toDouble();
+  mEdgeWidthUnit = QgsUnitTypes::decodeRenderUnit( QStringLiteral( "edge-width-unit" ) );
```
Comment (wonder-sk) [inline src/core/mesh/qgsmeshrenderersettings.cpp]: missing elem.attribute(...)

File: src/core/providers/meshmemory/qgsmeshmemorydataprovider.cpp [hunk 4]
```diff
@@ -290,12 +308,38 @@ bool QgsMeshMemoryDataProvider::checkDatasetValidity( std::shared_ptr<QgsMeshMem
...
+      setError( QgsError( tr( "Dataset defined on edges has {} values, but mesh {}" ).arg( dataset->values.count(), edgeCount() ),
+                          QStringLiteral( "Mesh Memory Provider" ) ) );
+    }
   }
-
   dataset->valid = valid;
   return valid;
 }
 
+bool QgsMeshMemoryDataProvider::checkVertexId( int vertex_id )
```
Comment (wonder-sk) [inline src/core/providers/meshmemory/qgsmeshmemorydataprovider.cpp]: vertex_id is [1..N] or [0..N-1] ? I think in either case one of the two followng ifs will need changing to <=

Comment (PeterPetrik) [reply src/core/providers/meshmemory/qgsmeshmemorydataprovider.cpp]: its [0..N-1], fixed <=

**Comment by pigreco** [comment]
Please add `Mesh` label

**Pull request merged**